### PR TITLE
[KYUUBI #5375] JDBC Engine supports PostgreSQL

### LINKED ISSUE #5375: [TASK][EASY] JDBC Engine supports PostgreSQL
### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### Search before creating

- [X] I have searched in the [task list](https://github.com/orgs/apache/projects/296) and found no similar tasks.

### Mentor

- [ ] I have sufficient knowledge and experience of this task, and I volunteer to be the mentor of this task to guide contributors to complete the task.

### Skill requirements

- Basic knowledge of Scala programming language
- Basic knowledge of JDBC concepts
- Basic knowledge of Docker and [Testcontianers](https://java.testcontainers.org/)

### Background and Goals

Currently, Kyuubi supports JDBC engines with limited dialects, e.g. Doris, Phoenix, we can easily extend the dialects to support more databases.

Reference https://www.bilibili.com/video/BV1UT411m7LR

### Implementation steps

- Implement a ClickHouse dialect in the JDBC engine
- Leverage the Testcontainers to add integration tests
- Update document correspondingly

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by ymZhao1001**
### _Why are the changes needed?_

Close #5375 

### _How was this patch tested?_
- [x] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [x] [Run test](https://kyuubi.readthedocs.io/en/master/contributing/code/testing.html#running-tests) locally before make a pull request

### _Was this patch authored or co-authored using generative AI tooling?_

No

### CODE REVIEW

**Comment by bowenliang123** [comment]
Run `dev/reformat` to solve the styling problems before committing.
https://kyuubi.readthedocs.io/en/master/contributing/code/developer.html#format-all-code

File: externals/kyuubi-jdbc-engine/src/main/scala/org/apache/kyuubi/engine/jdbc/dialect/PostgreSqlDialect.scala [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+    statement.setFetchSize(fetchSize)
+    statement
+  }
+
+  override def getTypeInfoOperation(session: Session): Operation = {
+    throw KyuubiSQLException.featureNotSupported()
+  }
+
+  override def getCatalogsOperation(): String = {
+    val query = new StringBuilder(
```
[the remark this answers is not in the document]
Comment (ymZhao1001) [reply externals/kyuubi-jdbc-engine/src/main/scala/org/apache/kyuubi/engine/jdbc/dialect/PostgreSqlDialect.scala]: updated

File: externals/kyuubi-jdbc-engine/src/main/scala/org/apache/kyuubi/engine/jdbc/dialect/PostgreSqlDialect.scala [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+import org.apache.kyuubi.operation.Operation
+import org.apache.kyuubi.operation.meta.ResultSetSchemaConstant._
+import org.apache.kyuubi.session.Session
+
+class PostgreSqlDialect extends JdbcDialect {
+
+  override def createStatement(connection: Connection, fetchSize: Int): Statement = {
+    val statement =
+      connection.createStatement(ResultSet.TYPE_FORWARD_ONLY, ResultSet.CONCUR_READ_ONLY)
+    statement.setFetchSize(fetchSize)
```
[the remark this answers is not in the document]
Comment (ymZhao1001) [reply externals/kyuubi-jdbc-engine/src/main/scala/org/apache/kyuubi/engine/jdbc/dialect/PostgreSqlDialect.scala:39]: connection.setAutoCommit(false)?

[the remark this answers is not in the document]
Comment (zhaomin1423) [reply externals/kyuubi-jdbc-engine/src/main/scala/org/apache/kyuubi/engine/jdbc/dialect/PostgreSqlDialect.scala:39]: Yes，and to commit after executing sql if auto commit is false.

File: externals/kyuubi-jdbc-engine/src/main/scala/org/apache/kyuubi/engine/jdbc/dialect/PostgreSqlDialect.scala [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+import org.apache.commons.lang3.StringUtils
+
+import org.apache.kyuubi.KyuubiSQLException
+import org.apache.kyuubi.engine.jdbc.postgresql.{PostgreSqlRowSetHelper, PostgreSqlSchemaHelper}
+import org.apache.kyuubi.engine.jdbc.schema.{RowSetHelper, SchemaHelper}
+import org.apache.kyuubi.operation.Operation
+import org.apache.kyuubi.operation.meta.ResultSetSchemaConstant._
+import org.apache.kyuubi.session.Session
+
+class PostgreSqlDialect extends JdbcDialect {
```
Comment (pan3793) [inline externals/kyuubi-jdbc-engine/src/main/scala/org/apache/kyuubi/engine/jdbc/dialect/PostgreSqlDialect.scala:34]: let's respect the original name of PostgreSQL

```suggestion
class PostgreSQLDialect extends JdbcDialect {
```

File: externals/kyuubi-jdbc-engine/src/main/scala/org/apache/kyuubi/engine/jdbc/dialect/PostgreSqlDialect.scala [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+         |FROM INFORMATION_SCHEMA.SCHEMATA
+         |""".stripMargin)
+
+    val filters = ArrayBuffer[String]()
+    if (StringUtils.isNotBlank(catalog)) {
+      filters += s"CATALOG_NAME = '$catalog'"
+    }
+
+    if (StringUtils.isNotBlank(schema)) {
+      filters += s"SCHEMA_NAME LIKE '$schema'"
```
Comment (pan3793) [inline externals/kyuubi-jdbc-engine/src/main/scala/org/apache/kyuubi/engine/jdbc/dialect/PostgreSqlDialect.scala:65]: `= $catalog` but `LIKE '$schema'`? the variable names is ambiguous

File: externals/kyuubi-jdbc-engine/src/main/scala/org/apache/kyuubi/engine/jdbc/dialect/PostgreSqlDialect.scala [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+      query.append(filters.mkString(" AND "))
+    }
+
+    query.toString()
+  }
+
+  override def getTablesQuery(
+      catalog: String,
+      schema: String,
+      tableName: String,
```
Comment (pan3793) [inline externals/kyuubi-jdbc-engine/src/main/scala/org/apache/kyuubi/engine/jdbc/dialect/PostgreSqlDialect.scala:79]: ditto, names are ambiguous

File: externals/kyuubi-jdbc-engine/src/main/resources/META-INF/services/org.apache.kyuubi.engine.jdbc.connection.JdbcConnectionProvider [hunk 6]
```diff
@@ -16,4 +16,5 @@
 #
 
 org.apache.kyuubi.engine.jdbc.doris.DorisConnectionProvider
-org.apache.kyuubi.engine.jdbc.phoenix.PhoenixConnectionProvider
\ No newline at end of file
+org.apache.kyuubi.engine.jdbc.phoenix.PhoenixConnectionProvider
+org.apache.kyuubi.engine.jdbc.postgresql.PostgreSqlConnectionProvider
```
Comment (pan3793) [inline externals/kyuubi-jdbc-engine/src/main/resources/META-INF/services/org.apache.kyuubi.engine.jdbc.connection.JdbcConnectionProvider:20]: let's add an additional newline at the end of the file to suppress the GitHub style complaint

**Comment by pan3793** [comment]
@ymZhao1001 good job, the PR is basically in good shape, I left a few comments, and don't forget to update the docs to mention this new feature.

File: externals/kyuubi-jdbc-engine/src/main/scala/org/apache/kyuubi/engine/jdbc/dialect/PostgreSqlDialect.scala (hunk 3 above)
Comment (ymZhao1001) [reply externals/kyuubi-jdbc-engine/src/main/scala/org/apache/kyuubi/engine/jdbc/dialect/PostgreSqlDialect.scala]: ok

File: externals/kyuubi-jdbc-engine/src/main/scala/org/apache/kyuubi/engine/jdbc/dialect/PostgreSqlDialect.scala (hunk 4 above)
Comment (ymZhao1001) [reply externals/kyuubi-jdbc-engine/src/main/scala/org/apache/kyuubi/engine/jdbc/dialect/PostgreSqlDialect.scala]: update

File: externals/kyuubi-jdbc-engine/src/main/scala/org/apache/kyuubi/engine/jdbc/dialect/PostgreSqlDialect.scala (hunk 5 above)
Comment (ymZhao1001) [reply externals/kyuubi-jdbc-engine/src/main/scala/org/apache/kyuubi/engine/jdbc/dialect/PostgreSqlDialect.scala]: update

File: externals/kyuubi-jdbc-engine/src/main/resources/META-INF/services/org.apache.kyuubi.engine.jdbc.connection.JdbcConnectionProvider (hunk 6 above)
Comment (ymZhao1001) [reply externals/kyuubi-jdbc-engine/src/main/resources/META-INF/services/org.apache.kyuubi.engine.jdbc.connection.JdbcConnectionProvider]: update

**Review by pan3793 — APPROVED** [review]

**Comment by pan3793** [comment]
Thanks, merged to master
